kong/clustering.lua

```diff
@@ -304,68 +326,117 @@ function _M.handle_cp_websocket()
   end
 
   -- connection established
-  -- ping thread
-  ngx.thread.spawn(function()
+  local ping_thread = ngx.thread.spawn(function()
+    local last_seen = ngx_time()
     while not exiting() do
       local data, typ, err = wb:recv_frame()
-      if not data then
-        ngx_log(ngx_ERR, "did not receive ping frame from data plane: ", err)
-        return ngx_exit(ngx_ERR)
+
+      if exiting() then
+        return
       end
 
-      assert(typ == "ping")
+      if err then
+        if not is_timeout(err) then
+          return nil, err
+        end
 
-      -- queue PONG to avoid races
-      table_insert(queue, "PONG")
-      queue.sem:post()
+        local waited = ngx_time() - last_seen
+        if waited > PING_WAIT then
+          return nil, "did not receive ping frame from data plane within " .. PING_WAIT .. " seconds"
+        end
+
+      else
+        if typ == "close" then
+          return
+        end
+
+        if not data then
+          return nil, "did not receive ping frame from data plane"
+        end
+
+        -- dps only send pings
+        if typ ~= "ping" then
+          return nil, "invalid websocket frame received from a data plane: " .. typ
+        end
 
-      local ok
-      ok, err = kong.db.clustering_data_planes:upsert({ id = node_id, }, {
-        last_seen = ngx_time(),
-        config_hash =
-          data ~= "" and data or nil,
-        hostname = node_hostname,
-        ip = node_ip,
-      }, { ttl = kong.configuration.cluster_data_plane_purge_delay, })
-      if not ok then
-        ngx_log(ngx_ERR, "unable to update clustering data plane status: ", err)
+        -- queue PONG to avoid races
+        table_insert(queue, "PONG")
+        queue.sem:post()
+
+        last_seen = ngx_time()
+
+        local ok
+        ok, err = kong.db.clustering_data_planes:upsert({ id = node_id, }, {
+          last_seen = last_seen,
+          config_hash = data ~= "" and data or nil,
+          hostname = node_hostname,
+          ip = node_ip,
+        }, { ttl = kong.configuration.cluster_data_plane_purge_delay, })
+        if not ok then
+          ngx_log(ngx_ERR, "unable to update clustering data plane status: ", err)
+        end
       end
     end
   end)
 
-  while not exiting() do
-    local ok, err = sem:wait(5)
-    if exiting() then
-      return
-    end
-    if ok then
-      local payload = table_remove(queue, 1)
-      assert(payload, "config queue can not be empty after semaphore returns")
-
-      if payload == "PONG" then
-        local _
-        _, err = wb:send_pong()
-        if err then
-          ngx_log(ngx_ERR, "failed to send PONG back to data plane: ", err)
-          return ngx_exit(ngx_ERR)
+  local main_thread = ngx.thread.spawn(function()
```

Comment:
Maybe call this `send_thread` and the `ping_thread` -> `read_thread` instead, it may makes the code easier to follow for us.